### PR TITLE
[PM-1879] Allow custom users to grant the same custom permissions that they have

### PR DESCRIPTION
**Pull request by r-tome**
## Type of change

```
- [ ] Bug fix
- [X] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective
The Manage Users custom permission should be updated to allow the custom user to grant the same custom permissions that they have, but not other permissions.

## Code changes
- **src/Api/Models/Request/Organizations/OrganizationUserRequestModels.cs:** Replaced `JsonSerializer.Serialize` with `CoreHelpers.ClassToJsonData`
- **src/Core/Services/Implementations/OrganizationService.cs:** Checking saving user's custom permissions
- **test/Core.Test/Services/OrganizationServiceTests.cs:** Added unit tests for saving custom permissions using a Custom saving user

## Before you submit

- Please check for formatting errors (`dotnet format --verify-no-changes`) (required)
- If making database changes - make sure you also update Entity Framework queries and/or migrations
- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

File: src/Core/Services/Implementations/OrganizationService.cs [hunk 1]
```diff
@@ -2115,9 +2117,9 @@ private void ValidateOrganizationUpgradeParameters(Models.StaticStore.Plan plan,
             return;
         }
 
-        if (oldType == OrganizationUserType.Custom || newType == OrganizationUserType.Custom)
+        if ((oldType == OrganizationUserType.Custom || newType == OrganizationUserType.Custom) && !await _currentContext.OrganizationCustom(organizationId))
         {
-            throw new BadRequestException("Only Owners and Admins can configure Custom accounts.");
+            throw new BadRequestException("Only Owners and Admins or other Custom users can configure Custom accounts.");
         }
```
Comment (eliykat) [inline src/Core/Services/Implementations/OrganizationService.cs:2123]: The purpose of this check is to prevent Custom users from modifying other Custom users. We now allow that in some circumstances, so I assume this check should be deleted and replaced with a call to `ValidateCustomPermissions` instead.

That also ensures that `ValidateCustomPermissions` is called from all the code flows that call this private method. There's quite a few similar ones in this service (save, invite, updateUserGroups, etc) and it doesn't look like they're all covered by the `ValidateCustomPermissions` check at the moment.

File: test/Core.Test/Services/OrganizationServiceTests.cs [hunk 2]
```diff
@@ -649,6 +648,78 @@ await sutProvider.GetDependency<IMailService>().Received(1)
         await sutProvider.Sut.SaveUserAsync(newUserData, savingUser.UserId, collections, groups);
     }
 
+    [Theory, BitAutoData]
+    public async Task SaveUser_WithCustomPermission_WhenCustomSavingUserHasPermission_Passes(
```
Comment (eliykat) [inline test/Core.Test/Services/OrganizationServiceTests.cs:652]: Slightly easier to read:

```suggestion
    public async Task SaveUser_WithCustomPermission_WhenSavingUserHasCustomPermission_Passes(
```

File: test/Core.Test/Services/OrganizationServiceTests.cs [hunk 3]
```diff
@@ -649,6 +648,78 @@ await sutProvider.GetDependency<IMailService>().Received(1)
...
+        Organization organization,
+        [OrganizationUser(type: OrganizationUserType.User)] OrganizationUser oldUserData,
+        [OrganizationUser(type: OrganizationUserType.Custom)] OrganizationUser newUserData,
+        IEnumerable<CollectionAccessSelection> collections,
+        IEnumerable<Guid> groups,
+        [OrganizationUser(type: OrganizationUserType.Custom)] OrganizationUser savingUser,
```
Comment (eliykat) [inline test/Core.Test/Services/OrganizationServiceTests.cs:658]: You also need to make the `savingUser` a custom user type, otherwise you might hit the early returns (if admin or owner) and get a false positive. (Assuming that the user role is random, maybe it defaults to something sensible)

File: test/Core.Test/Services/OrganizationServiceTests.cs [hunk 4]
```diff
@@ -649,6 +648,78 @@ await sutProvider.GetDependency<IMailService>().Received(1)
...
+            .Returns(new List<OrganizationUser> { savingUser });
+        currentContext.OrganizationCustom(savingUser.OrganizationId).Returns(true);
+        currentContext.ManageUsers(savingUser.OrganizationId).Returns(true);
+        currentContext.AccessReports(savingUser.OrganizationId).Returns(true);
+
+        await sutProvider.Sut.SaveUserAsync(newUserData, savingUser.UserId, collections, groups);
+    }
+
+    [Theory, BitAutoData]
+    public async Task SaveUser_WithCustomPermission_WhenCustomSavingUserDoesNotHavePermission_Throws(
```
Comment (eliykat) [inline test/Core.Test/Services/OrganizationServiceTests.cs:687]: As above

```suggestion
    public async Task SaveUser_WithCustomPermission_WhenSavingUserDoesNotHaveCustomPermission_Throws(
```

**Review by eliykat — CHANGES_REQUESTED** [review]

**Comment by eliykat** [comment]
This will also allow custom users to remove and revoke (and do anything else to) another custom user. It's not just giving permissions, it's also taking away. Please check with product that that's OK, as we only discussed inviting users. (I think it's OK I just want to make sure they know.)

Can you also double check there's no accidental privilege escalation here, like can a custom user now upgrade themselves to an admin? Please just try out some edge cases to make sure there's nothing we've missed.

File: src/Core/Services/Implementations/OrganizationService.cs [hunk 5]
```diff
@@ -2150,6 +2152,50 @@ private async Task ValidateOrganizationCustomPermissionsEnabledAsync(Guid organi
...
     }
 
+    private async Task ValidateCustomPermissions(OrganizationUser user)
```
Comment (eliykat) [inline src/Core/Services/Implementations/OrganizationService.cs:2155]: This method is only relevant if `user.Type == OrganizationUserType.Custom`, so it should either have an early return in this case, or only be called when required. (I personally prefer only calling methods when they're needed)

Comment (r-tome) [reply src/Core/Services/Implementations/OrganizationService.cs]: The method is now called on `ValidateOrganizationUserUpdatePermissions` when `newType == OrganizationUserType.Custom`

File: src/Core/Services/Implementations/OrganizationService.cs (hunk 1 above)
Comment (r-tome) [reply src/Core/Services/Implementations/OrganizationService.cs]: Good point, I modified `ValidateOrganizationUserUpdatePermissions` to call `ValidateCustomPermissionsGrant`

**Comment by r-tome** [comment]
> This will also allow custom users to remove and revoke (and do anything else to) another custom user. It's not just giving permissions, it's also taking away. Please check with product that that's OK, as we only discussed inviting users. (I think it's OK I just want to make sure they know.)

I've asked this, I'll let you know when I get a response.

> Can you also double check there's no accidental privilege escalation here, like can a custom user now upgrade themselves to an admin? Please just try out some edge cases to make sure there's nothing we've missed. I can imagine that the code might have made some assumptions about what custom users can do which are no longer accurate.

I added a new unit test to cover that scenario (`SaveUser_WithCustomPermission_WhenUpgradingToAdmin_Throws`)

File: src/Core/Services/Implementations/OrganizationService.cs [hunk 6]
```diff
@@ -2152,48 +2149,84 @@ private async Task ValidateOrganizationCustomPermissionsEnabledAsync(Guid organi
         }
     }
 
-    private async Task ValidateCustomPermissions(OrganizationUser user)
+    private async Task<bool> ValidateCustomPermissionsGrant(Guid organizationId, Permissions permissions)
```
Comment (eliykat) [inline src/Core/Services/Implementations/OrganizationService.cs:2152]: The old interface here was better - I think you should go back to passing in the [whole object](https://refactoring.guru/preserve-whole-object) (`OrganizationUser`). I assume you did it to fit the interface of `ValidateOrganizationUserUpdatePermissions`, but that too could be refactored to take the whole object, which would also be an improvement.

File: test/Core.Test/Services/OrganizationServiceTests.cs (hunk 3 above)
Comment (eliykat) [reply test/Core.Test/Services/OrganizationServiceTests.cs:688]: oh I missed that, it's even in the part that I commented on 😅

File: src/Core/Services/Implementations/OrganizationService.cs (hunk 6 above)
Comment (r-tome) [reply src/Core/Services/Implementations/OrganizationService.cs:2152]: That is what I intended to do but on `InviteUsersAsync` we do not have the `OrganizationUser` objects yet when calling `ValidateOrganizationUserUpdatePermissions` - they're only created afterwards.

Comment (eliykat) [reply src/Core/Services/Implementations/OrganizationService.cs:2152]: That's a good point. I think this is fine for now then.

**Review by eliykat — APPROVED** [review]
